### PR TITLE
TDL-19742 Run discover mode if catalog is not given in sync mode

### PR DESCRIPTION
**Pull request by jtilala**
# Description of change
 - Run discover mode and generate catalog when the catalog.json is not given in sync mode.
 - Write a unit test to justify the above scenario.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_catalog_in_sync.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+from shutil import ExecError
```
Comment (savan-chovatiya) [inline tests/unittests/test_catalog_in_sync.py:1]: Remove this unused import

File: tests/unittests/test_catalog_in_sync.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@mock.patch("tap_google_sheets.GoogleClient.__enter__")
+@mock.patch("tap_google_sheets.sync")
+class TestCatalog(unittest.TestCase):
+    def test_catalog_is_given_in_sync(self, mocked_sync, mocked_google_client, mocked_parse_args, mocked_discover):
+        """
+        To verify that if catalog is given in sync mode then run with catalog file
+        """
+        MockedParseArgs.catalog = "test"
+        mocked_parse_args.return_value = MockedParseArgs
+        mocked_google_client.return_value = "test"
```
Comment (savan-chovatiya) [inline tests/unittests/test_catalog_in_sync.py:24]: Move these two lines at the @mock.patch decorator level.

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tests/unittests/test_catalog_in_sync.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        """
+        To verify that if catalog is not given in sync mode then run discover mode to generate catalog 
+        """
+        # mocking discover function
+        MockedParseArgs.catalog = ""
+        mocked_discover.return_value = "test"
+        mocked_parse_args.return_value = MockedParseArgs
+        mocked_google_client.return_value = "test"
+        main()
+        mocked_sync.assert_called_with(client="test", config=MockedParseArgs.config, catalog="test", state={})
```
[the remark this answers is not in the document]
Comment (jtilala) [reply tests/unittests/test_catalog_in_sync.py]: Added assertion to check whether or not sync function calls discover()

File: tests/unittests/test_catalog_in_sync.py (hunk 2 above)
Comment (jtilala) [reply tests/unittests/test_catalog_in_sync.py]: Added return_value at decorator level and removed it from the test function

**Review by savan-chovatiya — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Review by shantanu73 — APPROVED** [review]

**Review by cosimon — APPROVED** [review]

**Pull request merged**